backends/cuda-gen/ceed-cuda-gen-operator-build.cpp

```diff
@@ -23,358 +23,227 @@
-    CeedCheck(eval_mode_in != CEED_EVAL_NONE || eval_mode_out != CEED_EVAL_NONE, ceed, CEED_ERROR_BACKEND,
-              "Backend does not implement restriction only identity operators");
-  }
-
-  ostringstream code;
-
-  // TODO: put in a function?
-  // Add atomicAdd function for old NVidia architectures
-  CeedCallBackend(CeedGetData(ceed, &ceed_data));
-  CeedCallBackend(cudaGetDeviceProperties(&prop, ceed_data->device_id));
-  if ((prop.major < 6) && (CEED_SCALAR_TYPE != CEED_SCALAR_FP32)) {
-    char       *atomic_add_source;
-    const char *atomic_add_path;
-
-    CeedCallBackend(CeedGetJitAbsolutePath(ceed, "ceed/jit-source/cuda/cuda-atomic-add-fallback.h", &atomic_add_path));
-    CeedDebug256(ceed, CEED_DEBUG_COLOR_SUCCESS, "----- Loading Atomic Add Source -----\n");
-    CeedCallBackend(CeedLoadSourceToBuffer(ceed, atomic_add_path, &atomic_add_source));
-    code << atomic_add_source;
-    CeedCallBackend(CeedFree(&atomic_add_path));
-    CeedCallBackend(CeedFree(&atomic_add_source));
-  }
-
-  // Load basis source files
-  // TODO: generalize to accept different device functions?
-  {
-    char       *tensor_basis_kernel_source;
-    const char *tensor_basis_kernel_path;
-
-    CeedCallBackend(CeedGetJitAbsolutePath(ceed, "ceed/jit-source/cuda/cuda-shared-basis-tensor-templates.h", &tensor_basis_kernel_path));
-    CeedDebug256(ceed, CEED_DEBUG_COLOR_SUCCESS, "----- Loading Tensor Basis Kernel Source -----\n");
-    CeedCallBackend(CeedLoadSourceToBuffer(ceed, tensor_basis_kernel_path, &tensor_basis_kernel_source));
-    code << tensor_basis_kernel_source;
-    CeedCallBackend(CeedFree(&tensor_basis_kernel_path));
-    CeedCallBackend(CeedFree(&tensor_basis_kernel_source));
-  }
-  {
-    char       *cuda_gen_template_source;
-    const char *cuda_gen_template_path;
-
-    CeedCallBackend(CeedGetJitAbsolutePath(ceed, "ceed/jit-source/cuda/cuda-gen-templates.h", &cuda_gen_template_path));
-    CeedDebug256(ceed, CEED_DEBUG_COLOR_SUCCESS, "----- Loading Cuda-Gen Template Source -----\n");
-    CeedCallBackend(CeedLoadSourceToBuffer(ceed, cuda_gen_template_path, &cuda_gen_template_source));
-    code << cuda_gen_template_source;
-    CeedCallBackend(CeedFree(&cuda_gen_template_path));
-    CeedCallBackend(CeedFree(&cuda_gen_template_source));
-  }
-
-  // Get QFunction source and name
-  string qfunction_source(qf_data->qfunction_source);
-  string qfunction_name(qf_data->qfunction_name);
-  string operator_name;
-  operator_name = "CeedKernelCudaGenOperator_" + qfunction_name;
-
+static int CeedOperatorBuildKernelData_Cuda_gen(Ceed ceed, CeedInt num_input_fields, CeedOperatorField *op_input_fields,
+                                                CeedQFunctionField *qf_input_fields, CeedInt num_output_fields, CeedOperatorField *op_output_fields,
+                                                CeedQFunctionField *qf_output_fields, CeedInt *max_P_1d, CeedInt *Q_1d, CeedInt *dim, bool *is_tensor,
```

Comment:
Maybe a bit nitpicky (and something I'm inconsistent on, but trying to get better at):

It'd be nice to have the parameters that are meant to be arrays as `type_t paramter[]` instead of `type_t *paramter` to help distinguish the output scalars from input/output arrays. This could also be solved with a docstring, but I don't think we want to be docstringing every single internal function.

Reply:
This isn't the current style, so I'll leave it for this PR and open an issue to see if we want to do a pass to standardize to this